indico/modules/categories/controllers/display.py

```diff
@@ -252,10 +253,12 @@ def _process_args(self):
         after = self._parse_year_month(request.args.get('after'))
         if before is None and after is None:
             raise BadRequest('"before" or "after" parameter must be specified')
-        hidden_event_ids = {e.id for e in self.category.get_hidden_events(user=session.user)}
-        event_query = (Event.query.with_parent(self.category)
+        hidden_event_ids = {e.id for e in self.category.get_hidden_events(user=session.user, deep=self.is_flat)}
+        event_query_filter = (Event.category_chain_overlaps(self.category.id) if self.is_flat
```

Comment:
This one should also use `is_visible_in`. Maybe it's time to add util for this since we use this in multiple places now...

```python
def get_event_query_filter(category, is_flat=False, hidden_event_ids=None):
    criteria = [
        Event.is_visible_in(category.id) if is_flat else (Event.category_id == category.id),
        ~Event.is_deleted,
    ]
    if hidden_event_ids:
        criteria.append(Event.id.notin_(hidden_event_ids))
    return db.and_(*criteria)
```